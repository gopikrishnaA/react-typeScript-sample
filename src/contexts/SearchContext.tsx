import React, { createContext } from "react";

export const SearchContext = createContext<{
  query: string;
  updateQuery: React.Dispatch<React.SetStateAction<string>>;
}>({
  query: "",
  updateQuery: (q) => {}
});

export const useSearchContext = () => React.useContext(SearchContext);

export const SearchProvider: React.FC<{ children: React.ReactNode }> = ({
  children
}) => {
  const [query, updateQuery] = React.useState("pizza");

  return (
    <SearchContext.Provider value={{ query, updateQuery }}>
      {children}
    </SearchContext.Provider>
  );
};

export const SearchConsumer = SearchContext.Consumer;
