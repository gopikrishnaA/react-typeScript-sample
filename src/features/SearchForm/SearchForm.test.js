import React from "react";
import { shallow } from "enzyme";
import { Form } from "reactstrap";
import renderer from 'react-test-renderer';

import * as SearchContext from "../../contexts/SearchContext";
import { SearchForm } from "./SearchForm";

jest.useFakeTimers();

const event = (value) => ({
  preventDefault: jest.fn(),
  currentTarget: {
    value
  }
})

describe("SearchForm", () => {
  let wrapper;
  let useEffect;

  const mockUseEffect = () => {
    useEffect.mockImplementationOnce(f => f());
  };

  const contextValues = { updateQuery: jest.fn() };

  beforeEach(() => {
    useEffect = jest.spyOn(React, "useEffect");
    mockUseEffect();

    jest
      .spyOn(SearchContext, "useSearchContext")
      .mockImplementation(() => contextValues);

  });

  afterEach(() => {
    jest.clearAllMocks();
  })

  describe("initial load", () => {
    beforeEach(() => {
      // Cache original functionality
      const realUseState = React.useState

      // Stub the initial state
      const stubInitialState = '12'

      // Mock useState before rendering your component
      jest
        .spyOn(React, 'useState')
        .mockImplementationOnce(() => realUseState(stubInitialState));
      jest
        .spyOn(React, "useRef")
        .mockReturnValue({
          current: {
            value: '12'
          }
        });
      wrapper = shallow(<SearchForm />);
    });
    afterEach(() => {
      jest.runAllTimers();
    })
    it("Should render on initial load", () => {
      const tree = renderer
            .create(wrapper.find(Form))
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
  });

  describe("Form Onsubmit handler", () => {
    beforeEach(() => {
      wrapper = shallow(<SearchForm />);
    });

    it("Should click on Form onSubmit when value < 3", () => {
      wrapper.find('input').props().onChange(event('12'));
      wrapper.find(Form).props().onSubmit(event('12'));
      expect(wrapper.find('input').props().value).toBe('12');
      expect(wrapper.find('small').props().children).toBe('Need be at least 3 characters');
    });

    it("Should click on Form onSubmit when value > 3", () => {
      wrapper.find('input').props().onChange(event('12345'));
      wrapper.find(Form).props().onSubmit(event('12345'));
      expect(wrapper.find('input').props().value).toBe('12345');
    });
  });
});
