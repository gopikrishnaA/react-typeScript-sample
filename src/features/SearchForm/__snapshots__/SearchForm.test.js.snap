// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm initial load Should render on initial load 1`] = `
<form
  className="d-flex align-items-start ml-3"
  method="post"
  onSubmit={[Function]}
>
  <div
    className="d-flex flex-column justify-content-start search-form"
  >
    <input
      className="mr-2 search-input border-0"
      onChange={[Function]}
      placeholder="ex. pizza"
      type="text"
      value="12"
    />
     
    <i
      className="fas fa-search i-search"
      onClick={[Function]}
    />
  </div>
</form>
`;
